File: dbflow-processor/src/main/java/com/raizlabs/android/dbflow/processor/definition/column/ForeignKeyReferenceDefinition.java
package com.raizlabs.android.dbflow.processor.definition.column;

import com.raizlabs.android.dbflow.annotation.ForeignKeyReference;
import com.raizlabs.android.dbflow.processor.model.ProcessorManager;
import com.raizlabs.android.dbflow.processor.utils.ElementUtility;
import com.raizlabs.android.dbflow.processor.utils.ModelUtils;
import com.raizlabs.android.dbflow.sql.QueryBuilder;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.TypeName;

import java.util.concurrent.atomic.AtomicInteger;

import javax.lang.model.element.TypeElement;
import javax.lang.model.type.MirroredTypeException;
import javax.lang.model.type.TypeMirror;

/**
 * Description:
 */
public class ForeignKeyReferenceDefinition {

    private final ProcessorManager manager;
    private final String foreignKeyFieldName;

    public final String columnName;
    public final String foreignColumnName;
    public final TypeName columnClassName;

    private boolean isReferencedFieldPrivate;
    private boolean isReferencedFieldPackagePrivate;

    public BaseColumnAccess columnAccess;

    private final BaseColumnAccess tableColumnAccess;
    private final ForeignKeyColumnDefinition foreignKeyColumnDefinition;

    private final BaseColumnAccess simpleColumnAccess;

    public ForeignKeyReferenceDefinition(ProcessorManager manager, String foreignKeyFieldName,
                                         ColumnDefinition referencedColumn,
                                         BaseColumnAccess tableColumnAccess,
                                         ForeignKeyColumnDefinition foreignKeyColumnDefinition, int referenceCount) {
        this.manager = manager;
        this.foreignKeyColumnDefinition = foreignKeyColumnDefinition;
        this.tableColumnAccess = tableColumnAccess;
        this.foreignKeyFieldName = foreignKeyFieldName;

        if (!foreignKeyColumnDefinition.getIsPrimaryKey() && !foreignKeyColumnDefinition.isPrimaryKeyAutoIncrement() && !foreignKeyColumnDefinition.getIsRowId()
                || referenceCount > 0) {
            columnName = foreignKeyFieldName + "_" + referencedColumn.getColumnName();
        } else {
            columnName = foreignKeyFieldName;
        }
        foreignColumnName = referencedColumn.getColumnName();
        columnClassName = referencedColumn.elementTypeName;

        if (referencedColumn.getColumnAccess() instanceof WrapperColumnAccess) {
            isReferencedFieldPrivate = (((WrapperColumnAccess) referencedColumn.getColumnAccess()).existingColumnAccess instanceof PrivateColumnAccess);
            isReferencedFieldPackagePrivate = ((WrapperColumnAccess) referencedColumn.getColumnAccess()).existingColumnAccess instanceof PackagePrivateAccess;
        } else {
            isReferencedFieldPrivate = (referencedColumn.getColumnAccess() instanceof PrivateColumnAccess);

            // fix here to ensure we can access it otherwise we generate helper
            boolean isPackagePrivate = ElementUtility.isPackagePrivate(referencedColumn.element);
            boolean isPackagePrivateNotInSamePackage = isPackagePrivate && !ElementUtility.isInSamePackage(manager, referencedColumn.element, foreignKeyColumnDefinition.element);

            isReferencedFieldPackagePrivate = referencedColumn.getColumnAccess() instanceof PackagePrivateAccess
                    || isPackagePrivateNotInSamePackage;
        }
        if (isReferencedFieldPrivate) {
            columnAccess = new PrivateColumnAccess(referencedColumn.getColumn(), false);
        } else if (isReferencedFieldPackagePrivate) {
            columnAccess = new PackagePrivateAccess(referencedColumn.packageName,
                    foreignKeyColumnDefinition.tableDefinition.getDatabaseDefinition().classSeparator,
                    ClassName.get((TypeElement) referencedColumn.element.getEnclosingElement()).simpleName());
            PackagePrivateAccess.putElement(((PackagePrivateAccess) columnAccess).helperClassName, foreignColumnName);
        } else {
            columnAccess = new SimpleColumnAccess();
        }

        simpleColumnAccess = new SimpleColumnAccess(columnAccess instanceof PackagePrivateAccess);
    }

    public ForeignKeyReferenceDefinition(ProcessorManager manager, String foreignKeyFieldName,
                                         ForeignKeyReference foreignKeyReference, BaseColumnAccess tableColumnAccess,
                                         ForeignKeyColumnDefinition foreignKeyColumnDefinition) {
        this.manager = manager;
        this.tableColumnAccess = tableColumnAccess;
        this.foreignKeyColumnDefinition = foreignKeyColumnDefinition;
        this.foreignKeyFieldName = foreignKeyFieldName;

        columnName = foreignKeyReference.columnName();
        foreignColumnName = foreignKeyReference.foreignKeyColumnName();

        TypeMirror columnClass = null;
        try {
            foreignKeyReference.columnType();
        } catch (MirroredTypeException mte) {
            columnClass = mte.getTypeMirror();
        }
        columnClassName = TypeName.get(columnClass);
        isReferencedFieldPrivate = foreignKeyReference.referencedFieldIsPrivate();
        isReferencedFieldPackagePrivate = foreignKeyReference.referencedFieldIsPackagePrivate();
        if (isReferencedFieldPrivate) {
            columnAccess = new PrivateColumnAccess(foreignKeyReference);
        } else if (isReferencedFieldPackagePrivate) {
            columnAccess = new PackagePrivateAccess(foreignKeyColumnDefinition.referencedTableClassName.packageName(),
                    foreignKeyColumnDefinition.tableDefinition.getDatabaseDefinition().classSeparator,
                    foreignKeyColumnDefinition.referencedTableClassName.simpleName());
            PackagePrivateAccess.putElement(((PackagePrivateAccess) columnAccess).helperClassName, foreignColumnName);
        } else {
            columnAccess = new SimpleColumnAccess();
        }

        simpleColumnAccess = new SimpleColumnAccess(columnAccess instanceof PackagePrivateAccess);
    }

    CodeBlock getCreationStatement() {
        return DefinitionUtils.getCreationStatement(columnClassName, null, columnName).build();
    }

    String getPrimaryKeyName() {
        return QueryBuilder.quote(columnName);
    }

    CodeBlock getContentValuesStatement(boolean isModelContainerAdapter) {
        // fix its access here.
        CodeBlock shortAccess = tableColumnAccess.getShortAccessString(foreignKeyColumnDefinition.elementClassName, foreignKeyFieldName, false);
        shortAccess = foreignKeyColumnDefinition.getForeignKeyReferenceAccess(shortAccess);

        CodeBlock columnShortAccess = getShortColumnAccess(false, shortAccess);

        CodeBlock combined;
        if (!(columnAccess instanceof PackagePrivateAccess)) {
            combined = CodeBlock.of("$L$L$L", shortAccess, (isModelContainerAdapter ? "" : "."),
                    columnShortAccess);
        } else {
            combined = columnShortAccess;
        }
        return DefinitionUtils.getContentValuesStatement(columnShortAccess.toString(),
                combined.toString(),
                columnName, columnClassName, simpleColumnAccess,
                getForeignKeyColumnVariable(), null,
                foreignKeyColumnDefinition.tableDefinition.outputClassName).build();
    }

    public CodeBlock getPrimaryReferenceString() {
        CodeBlock shortAccess = tableColumnAccess.getShortAccessString(foreignKeyColumnDefinition.elementClassName, foreignKeyFieldName, false);
        shortAccess = foreignKeyColumnDefinition.getForeignKeyReferenceAccess(shortAccess);
        CodeBlock columnShortAccess = getShortColumnAccess(false, shortAccess);
        CodeBlock combined;
        if (!(columnAccess instanceof PackagePrivateAccess)) {
            combined = CodeBlock.of("$L.$L.$L", ModelUtils.getVariable(),
                    shortAccess, columnShortAccess);
        } else {
            combined = columnShortAccess;
        }
        return combined;
    }

    CodeBlock getSQLiteStatementMethod(AtomicInteger index) {
        CodeBlock shortAccess = tableColumnAccess.getShortAccessString(foreignKeyColumnDefinition.elementClassName, foreignKeyFieldName, true);
        shortAccess = foreignKeyColumnDefinition.getForeignKeyReferenceAccess(shortAccess);

        CodeBlock columnShortAccess = getShortColumnAccess(true, shortAccess);
        CodeBlock combined = shortAccess.toBuilder()
                .add(".")
                .add(columnShortAccess).build();
        return DefinitionUtils.getSQLiteStatementMethod(
                index, columnShortAccess.toString(), combined.toString(),
                columnClassName, simpleColumnAccess,
                getForeignKeyColumnVariable(), false, null).build();
    }

    CodeBlock getForeignKeyContainerMethod(String referenceFieldName, CodeBlock loadFromCursorBlock) {

        CodeBlock codeBlock = columnAccess
                .setColumnAccessString(columnClassName, foreignColumnName, foreignColumnName,
                        referenceFieldName, loadFromCursorBlock);
        CodeBlock.Builder codeBuilder = CodeBlock.builder();
        codeBuilder.addStatement("$L", codeBlock);
        return codeBuilder.build();
    }

    private String getForeignKeyColumnVariable() {
        return ModelUtils.getVariable();
    }

    private CodeBlock getShortColumnAccess(boolean isSqliteMethod, CodeBlock shortAccess) {
        if (columnAccess instanceof PackagePrivateAccess) {
            return columnAccess.getColumnAccessString(columnClassName, foreignColumnName, "", ModelUtils.getVariable() + "." + shortAccess, isSqliteMethod);
        } else {
            return columnAccess.getShortAccessString(columnClassName, foreignColumnName, isSqliteMethod);
        }
    }

}

File: dbflow-processor/src/main/java/com/raizlabs/android/dbflow/processor/definition/column/ForeignKeyReferenceDefinition.kt
package com.raizlabs.android.dbflow.processor.definition.column

import com.raizlabs.android.dbflow.data.Blob
import com.raizlabs.android.dbflow.processor.ProcessorManager
import com.raizlabs.android.dbflow.processor.definition.TypeConverterDefinition
import com.raizlabs.android.dbflow.processor.utils.ElementUtility
import com.raizlabs.android.dbflow.processor.utils.isNullOrEmpty
import com.raizlabs.android.dbflow.sql.QueryBuilder
import com.squareup.javapoet.ClassName
import com.squareup.javapoet.CodeBlock
import com.squareup.javapoet.TypeName
import javax.lang.model.element.TypeElement

/**
 * Description:
 */
class ForeignKeyReferenceDefinition(private val manager: ProcessorManager,
                                    foreignKeyFieldName: String,
                                    foreignKeyElementName: String, referencedColumn: ColumnDefinition,
                                    private val foreignKeyColumnDefinition: ForeignKeyColumnDefinition,
                                    referenceCount: Int, localColumnName: String = "") {

    val columnName: String
    val foreignColumnName: String
    val columnClassName: TypeName?

    var hasTypeConverter: Boolean = false

    internal val creationStatement: CodeBlock
        get() = DefinitionUtils.getCreationStatement(columnClassName, wrapperTypeName, columnName).build()

    internal val primaryKeyName: String
        get() = QueryBuilder.quote(columnName)

    private var isReferencedFieldPrivate: Boolean = false
    private var isReferencedFieldPackagePrivate: Boolean = false

    lateinit var columnAccessor: ColumnAccessor
    var wrapperAccessor: ColumnAccessor? = null
    var wrapperTypeName: TypeName? = null
    var subWrapperAccessor: ColumnAccessor? = null

    lateinit var partialAccessor: PartialLoadFromCursorAccessCombiner
    lateinit var primaryReferenceField: ForeignKeyAccessField
    lateinit var contentValuesField: ForeignKeyAccessField
    lateinit var sqliteStatementField: ForeignKeyAccessField

    private fun createScopes(foreignKeyColumnDefinition: ForeignKeyColumnDefinition,
                             foreignKeyFieldName: String, getterSetter: GetterSetter,
                             name: String, packageName: String) {
        if (isReferencedFieldPrivate) {
            columnAccessor = PrivateScopeColumnAccessor(foreignKeyFieldName, getterSetter, false)
        } else if (isReferencedFieldPackagePrivate) {
            columnAccessor = PackagePrivateScopeColumnAccessor(foreignKeyFieldName, packageName,
                    foreignKeyColumnDefinition.baseTableDefinition.databaseDefinition?.classSeparator,
                    name)

            PackagePrivateScopeColumnAccessor.putElement(
                    (columnAccessor as PackagePrivateScopeColumnAccessor).helperClassName,
                    foreignKeyFieldName)
        } else {
            columnAccessor = VisibleScopeColumnAccessor(foreignKeyFieldName)
        }
    }


    private fun createForeignKeyFields(columnClassName: TypeName?, foreignKeyColumnDefinition: ForeignKeyColumnDefinition, manager: ProcessorManager) {
        val typeConverterDefinition = columnClassName?.let { manager.getTypeConverterDefinition(it) }
        evaluateTypeConverter(typeConverterDefinition)

        val combiner = Combiner(columnAccessor, columnClassName!!, wrapperAccessor,
                wrapperTypeName, subWrapperAccessor)
        partialAccessor = PartialLoadFromCursorAccessCombiner(columnName, foreignColumnName,
                columnClassName, foreignKeyColumnDefinition.baseTableDefinition.orderedCursorLookUp,
                columnAccessor, wrapperAccessor, wrapperTypeName)

        primaryReferenceField = ForeignKeyAccessField(columnName, PrimaryReferenceAccessCombiner(combiner))

        contentValuesField = ForeignKeyAccessField(columnName, ContentValuesCombiner(combiner))

        sqliteStatementField = ForeignKeyAccessField("start", SqliteStatementAccessCombiner(combiner))

    }

    private fun evaluateTypeConverter(typeConverterDefinition: TypeConverterDefinition?) {
        // Any annotated members, otherwise we will use the scanner to find other ones
        typeConverterDefinition?.let {

            if (it.modelTypeName != columnClassName) {
                manager.logError("The specified custom TypeConverter's Model Value %1s from %1s must match the type of the column %1s. ",
                        it.modelTypeName, it.className, columnClassName)
            } else {
                hasTypeConverter = true

                val fieldName = foreignKeyColumnDefinition.baseTableDefinition
                        .addColumnForTypeConverter(foreignKeyColumnDefinition, it.className)
                wrapperAccessor = TypeConverterScopeColumnAccessor(fieldName)
                wrapperTypeName = it.dbTypeName

                // special case of blob
                if (wrapperTypeName == ClassName.get(Blob::class.java)) {
                    subWrapperAccessor = BlobColumnAccessor()
                }
            }
        }
    }

    init {
        if (!localColumnName.isNullOrEmpty()) {
            this.columnName = localColumnName
        } else if (!foreignKeyColumnDefinition.isPrimaryKey && !foreignKeyColumnDefinition.isPrimaryKeyAutoIncrement
                && !foreignKeyColumnDefinition.isRowId || referenceCount > 0) {
            this.columnName = foreignKeyFieldName + "_" + referencedColumn.columnName
        } else {
            this.columnName = foreignKeyFieldName
        }
        foreignColumnName = referencedColumn.columnName
        this.columnClassName = referencedColumn.elementTypeName
        isReferencedFieldPrivate = referencedColumn.columnAccessor is PrivateScopeColumnAccessor
        isReferencedFieldPackagePrivate = referencedColumn.columnAccessor is PackagePrivateScopeColumnAccessor
        val isPackagePrivate = ElementUtility.isPackagePrivate(referencedColumn.element)
        val isPackagePrivateNotInSamePackage = isPackagePrivate &&
                !ElementUtility.isInSamePackage(manager, referencedColumn.element,
                        foreignKeyColumnDefinition.element)
        isReferencedFieldPackagePrivate = isReferencedFieldPackagePrivate || isPackagePrivateNotInSamePackage
        val packageName = referencedColumn.packageName
        val name = ClassName.get(referencedColumn.element.enclosingElement as TypeElement).simpleName()
        createScopes(foreignKeyColumnDefinition, foreignKeyElementName, object : GetterSetter {
            override val getterName: String = referencedColumn.column?.getterName ?: ""
            override val setterName: String = referencedColumn.column?.setterName ?: ""
        }, name, packageName)
        createForeignKeyFields(columnClassName, foreignKeyColumnDefinition, manager)
    }


}
